### PR TITLE
Split plugin

### LINKED ISSUE #293: Feature: Split array of objects into Events
New plugin to split array of objects into Events

### PR DESCRIPTION
**Pull request by vadimalekseev**
Fixes #293

### CODE REVIEW

File: pipeline/batch.go [hunk 1]
```diff
@@ -11,7 +11,9 @@ import (
 )
 
 type Batch struct {
-	Events []*Event
+	events        []*Event
+	iteratorIndex int
```
Comment (vano144) [inline pipeline/batch.go:15]: let's add comment about what it is

File: pipeline/processor.go [hunk 2]
```diff
@@ -197,11 +199,16 @@ func (p *processor) doActions(event *Event) (isPassed bool) {
...
-		switch action.Do(event) {
-		case ActionPass:
+		result := action.Do(event)
+		switch result {
+		case ActionPass, ActionSpawned:
 			p.countEvent(event, index, eventStatusPassed)
 			p.tryResetBusy(index)
 			p.actionWatcher.setEventAfter(index, event, eventStatusPassed)
+
+			if result == ActionSpawned {
```
Comment (vano144) [inline pipeline/processor.go:209]: is it useless ?

Comment (vadimalekseev) [reply pipeline/processor.go:209]: In `ActionPass` case, we will continue process this (Parent) event in the next actions :)

My idea was that we should not continue to process parent events

File: pipeline/batch.go (hunk 1 above)
Comment (vadimalekseev) [reply pipeline/batch.go:16]: done

**Review by kirillov6 — APPROVED** [review]

**Review by HeadHunter483 — APPROVED** [review]

File: pipeline/processor.go [hunk 3]
```diff
@@ -10,17 +11,18 @@ type ActionResult int
...
 	// ActionDiscard skip further processing of event and request next event from any stream and source
-	ActionDiscard ActionResult = 1
+	ActionDiscard
 	// ActionHold hold event in a plugin and request next event from the same stream and source as current.
 	// same as ActionCollapse but held event should be manually committed or returned into pipeline.
 	// check out Commit()/Propagate() functions in InputPluginController.
 	// plugin may receive event with EventKindTimeout if it takes to long to read next event from same stream.
-	ActionHold ActionResult = 3
+	ActionHold
+	ActionSpawned = 4
```
Comment (vitkovskii) [inline pipeline/processor.go:25]: Action should be a verb. `Spawned` isn't a verb.

Comment (vitkovskii) [reply pipeline/processor.go:25]: Write a comment please about new action

**Review by vitkovskii — CHANGES_REQUESTED** [review]

File: pipeline/processor.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vadimalekseev) [reply pipeline/processor.go]: done

Comment (vadimalekseev) [reply pipeline/processor.go]: Done. Now its `ActionBreak`

File: plugin/action/remap/remap.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	p.pluginController = params.Controller
+}
+
+func (p *Plugin) Stop() {
+}
+
+func (p *Plugin) Do(event *pipeline.Event) pipeline.ActionResult {
+	node := event.Root.Dig(p.config.Field_...)
+	if !node.IsArray() {
+		p.logger.Warn("skip an event because is not an array", zap.String("type", node.TypeStr()))
```
[the remark this answers is not in the document]
Comment (vadimalekseev) [reply plugin/action/remap/remap.go]: done

**Review by vano144 — APPROVED** [review]

File: pipeline/processor.go [hunk 5]
```diff
@@ -341,6 +351,41 @@ func (p *processor) Propagate(event *Event) {
...
+// Spawn the children of the parent and process in the actions.
+// Any attempts to ActionHold or ActionCollapse the event will be suppressed by timeout events.
+func (p *processor) Spawn(parent *Event, nodes []*insaneJSON.Node) {
+	parent.SetChildParentKind()
+	nextActionIdx := parent.action + 1
+
+	for _, node := range nodes {
+		child := newEvent()
+		parent.children = append(parent.children, child)
+		child.Root.MutateToNode(node)
```
Comment (vano144) [inline pipeline/processor.go:363]: let's add comment about it

Comment (vadimalekseev) [reply pipeline/processor.go:365]: done

**Pull request merged**